### PR TITLE
Stop using custom OrderedDict

### LINKED ISSUE #3185: deprecated import of Iterable
Another tracking/maintenance issue. From the python 3.7 Travis CI build:
```
/home/travis/build/SCons/scons/src/engine/SCons/Util.py:49: DeprecationWarning: Using or importing the ABCs from 'collections' instead of from 'collections.abc' is deprecated, and in 3.8 it will stop working
  from collections import Iterable
```
From Python documentation:
https://docs.python.org/3/library/collections.html#module-collections

> Changed in version 3.3: Moved Collections Abstract Base Classes to the collections.abc module. For backwards compatibility, they continue to be visible in this module through Python 3.7. Subsequently, they will be removed entirely.

and
https://docs.python.org/3/library/collections.abc.html

### PR DESCRIPTION
**Pull request by mwichmann**
OrdredDict is in the standard library for all supported Python versions
(2.7 and 3.5+) and has improvements over the ActiveState recipe version
of OrderedDict we have been using. Switch to importing from collections
instead of getting it from SCons.Util (tests already did this).

At the same time, reorganize the Util.py imports - import Iterable
from collections.abc if possible (it is deprecated to import
it from collections, will stop working in 3.8); try getting the
User{Dict,List,String} from collections if possible - that is, try the
3.x way first.  Fixes #3185.

Signed-off-by: Mats Wichmann <mats@linux.com>

## Contributor Checklist:

* [X] I have created a new test or updated the unit tests to cover the new/changed functionality.
* [X] I have updated `master/src/CHANGES.txt` directory (and read the `README.txt` in that directory)
* [X] I have updated the appropriate documentation

Tests referencing OrderedDict already imported the stdlib one from collections, so no impact. There is no doc mention of this type.

### CODE REVIEW

**Comment by bdbaddog** [comment]
Can you add a TODO: Remove ordereddict usage when we move the floor to python 3.5?
I beleive from that version forward there's a guarantee that insertion order is maintained?

Also does it make sense to swap the collections.abc import so the common /current usage is first and the exception handling is the least common case?

**Comment by mwichmann** [comment]
re: collections.abc: that was my intent, but it depends on what you consider "current", 2.7 or 3.x.  From Py3 docs: "Changed in version 3.3: Moved Collections Abstract Base Classes to the collections.abc module. For backwards compatibility, they continue to be visible in this module through Python 3.7. Subsequently, they will be removed entirely."  so importing from the abc file is current for all python version except 2.7.

dict ordering: I checked the details. cpython preserves order in 3.6 (not 3.5), but that's considered an "implementation detail" and not required by the language. It became a requirement in 3.7. So the floor is a little higher; I don't know if you are expected to stop using OrderedDict as of that point.

File: src/engine/SCons/Util.py [hunk 1]
```diff
@@ -63,7 +60,7 @@
 FunctionType    = types.FunctionType
 
 try:
-    unicode
+    _ = type(unicode)
```
Comment (bdbaddog) [inline src/engine/SCons/Util.py:63]: Why? (Curious)

Comment (mwichmann) [reply src/engine/SCons/Util.py:63]: just to be running a "real" Python command which still generates a NameError on Py3.  I don't have any particular attachment to this, PyCharm grumped at the original, but don't mind going back, this wasn't an important part of the change.
Py2:
```
>>> unicode
<type 'unicode'>
>>> _ = type(unicode)
>>> 
```
Py3:
```
>>> unicode
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
NameError: name 'unicode' is not defined
>>> _ = type(unicode)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
NameError: name 'unicode' is not defined
>>> 
```

Comment (bdbaddog) [reply src/engine/SCons/Util.py:63]: Interesting.
O.k. leave it as is.

At some point I'd like to add pylint  or equivalent to our CI..

**Comment by bdbaddog** [comment]
Can you update src/CHANGES.txt to say remove obsoleted internal implementaiton of OrderedDict.. and then I'll go ahead and merge.

**Pull request merged**